### PR TITLE
Add `reference` data store

### LINKED ISSUE #928: Efficiently access NetCDF files in S3
**Is your feature request related to a problem? Please describe.**

We need to efficiently access a huge archive of climate data files in NetCDF format stored in a S3-compatible object storage via the data store framework. The data files are indexed using [kerchunk](https://github.com/fsspec/kerchunk).

**Describe the solution you'd like**

Add a new filesystem data store that supports the [`reference` protocol](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.reference.ReferenceFileSystem). 

**Additional context**

https://climate.esa.int/

### LINKED ISSUE #930: `KeyError: 'lon_bnds'` raised occasionally when opening (mostly NetCDF) datasets
**Describe the bug**

We sometimes see `KeyError: 'lon_bnds'` raised when opening datasets, mostly NetCDF. Cause of the problem is that an (internal) xcube function `get_dataset_bounds()` expects bounds coordinates, e.g. `lon_bnds`, to always be in `xarray.Dataset.coords`. However, xarray often puts such coordinates in  `xarray.Dataset.data_vars`. The name of bounds coordinates are usually given by the value of a coordinate attribute named `bounds` - according to CF conventions.

**Expected behavior**

1. check if given bounds coordinate is contained in dataset
2. get bounds coordinate from `xarray.Dataset.variables`, not `xarray.Dataset.coords`

**Traceback**

```
...
File ~/projects/xcube/xcube/core/store/descriptor.py:432, in _determine_bbox(data)
    429 def _determine_bbox(data: xr.Dataset) \
    430         -> Optional[Tuple[float, float, float, float]]:
    431     try:
--> 432         return get_dataset_bounds(data)
    433     except ValueError:
    434         if 'geospatial_lon_min' in data.attrs and \
    435                 'geospatial_lat_min' in data.attrs and \
    436                 'geospatial_lon_max' in data.attrs and \
    437                 'geospatial_lat_max' in data.attrs:

File ~/projects/xcube/xcube/core/geom.py:726, in get_dataset_bounds(dataset, xy_var_names)
    724 x_bnds_name = get_dataset_bounds_var_name(dataset, x_name)
    725 if x_bnds_name:
--> 726     x_bnds_var = dataset.coords[x_bnds_name]
    727     x1 = x_bnds_var[0, 0]
    728     x2 = x_bnds_var[0, 1]

File ~/mamba/envs/xcube/Lib/site-packages/xarray/core/coordinates.py:581, in DatasetCoordinates.__getitem__(self, key)
    579 def __getitem__(self, key: Hashable) -> DataArray:
    580     if key in self._data.data_vars:
--> 581         raise KeyError(key)
    582     return self._data[key]

KeyError: 'lon_bnds'
```

### PR DESCRIPTION
**Pull request by forman**
Added new `reference` data store to support "kerchunked" NetCDF files in object storage. 

Closes #928
Closes #930

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in `docs/source/*`
* [x] Changes documented in `CHANGES.md`
* [ ] GitHub CI passes
* [ ] AppVeyor CI passes
* [ ] Test coverage remains or increases (target 100%)

### CODE REVIEW

File: xcube/core/geom.py [hunk 1]
```diff
@@ -722,8 +722,8 @@ def get_dataset_bounds(dataset: Union[xr.Dataset, xr.DataArray],
 
     # Note, x_min > x_max then we intersect with the anti-meridian
     x_bnds_name = get_dataset_bounds_var_name(dataset, x_name)
-    if x_bnds_name:
-        x_bnds_var = dataset.coords[x_bnds_name]
+    if x_bnds_name and x_bnds_name in dataset:
```
Comment (TonioF) [inline xcube/core/geom.py:725]: This seems redundant to me. x_bnds_name is derived from the dataset one line earlier. We know it is in dataset. The fix is the line below. Same comment for y_bnds_name, of course.

File: xcube/core/store/fs/impl/fs.py [hunk 2]
```diff
@@ -19,6 +19,7 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
+from xcube.util.jsonschema import JsonArraySchema
```
Comment (TonioF) [inline xcube/core/store/fs/impl/fs.py:22]: Why? This is never used.

File: xcube/core/store/ref/store.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import os
+import warnings
+from typing import Iterator, Any, Tuple, Container, Union, Dict, List, \
+    Optional
```
Comment (TonioF) [inline xcube/core/store/ref/store.py:25]:
```suggestion
from typing import Iterator, Any, Tuple, Container, Union, Dict, List
```

File: xcube/core/store/ref/store.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        self._refs = dict(self._normalize_ref(ref) for ref in refs)
+        self._ref_kwargs = ref_kwargs
+
+    @classmethod
+    def get_data_store_params_schema(cls) -> JsonObjectSchema:
+        return REF_STORE_SCHEMA
+
+    @classmethod
+    def get_data_types(cls) -> Tuple[str, ...]:
+        return ("dataset",)
```
Comment (TonioF) [inline xcube/core/store/ref/store.py:86]:
```suggestion
        return "dataset",
```

File: xcube/core/store/ref/store.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        if data_descriptor is None:
+            dataset = self.open_data(data_id)
+            data_descriptor = new_data_descriptor(data_id, dataset)
+            self._refs[data_id]["data_descriptor"] = data_descriptor
+        return data_descriptor
+
+    def get_data_opener_ids(self,
+                            data_id: str = None,
+                            data_type: DataTypeLike = None) -> Tuple[str, ...]:
+        return ("dataset:zarr:reference",)
```
Comment (TonioF) [inline xcube/core/store/ref/store.py:115]:
```suggestion
        return "dataset:zarr:reference",
```

**Review by TonioF — CHANGES_REQUESTED** [review]
I installed kerchunk to run the tests that wouldn't be executed otherwise. I found that a few one of them failed, because one of the sst values in the computed dataset were different in the second decimal place. I find this negligible for now, but we should understand why that is.

File: xcube/core/store/ref/store.py (hunk 5 above)
Comment (forman) [reply xcube/core/store/ref/store.py:115]: I disagree. This is intentional to make clear its a tuple and not a misplaced comma.

File: xcube/core/store/fs/impl/fs.py (hunk 2 above)
Comment (forman) [reply xcube/core/store/fs/impl/fs.py:22]: Leftover.

File: xcube/core/store/ref/store.py (hunk 3 above)
Comment (forman) [reply xcube/core/store/ref/store.py:25]: Ok.

File: xcube/core/store/ref/store.py (hunk 4 above)
Comment (forman) [reply xcube/core/store/ref/store.py:86]: I disagree. This is intentional to make clear its a tuple and not a misplaced comma.

**Comment by forman** [comment]
@TonioF thanks for reviewing. I also updated the unit-test mentioned.

File: test/core/store/ref/test_store.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        self.assertEqual({'time': 3, 'lat': 1350, 'lon': 1600}, cube.sizes)
+        self.assertEqual({'lon', 'time', 'lat'}, set(cube.coords))
+        self.assertEqual(
+            {'analysis_error', 'mask', 'analysed_sst', 'sea_ice_fraction'},
+            set(cube.data_vars))
+        sst_ts = cube.isel(lat=0, lon=0).compute()
+        sst_data = sst_ts.analysed_sst.values
+        self.assertAlmostEqual(290.02, sst_data[0])
+        self.assertAlmostEqual(289.94, sst_data[1])
+        self.assertAlmostEqual(289.89, sst_data[2])
```
Comment (TonioF) [inline test/core/store/ref/test_store.py:100]:
```suggestion
        self.assertAlmostEqual(290.02, sst_data[0], 2)
        self.assertAlmostEqual(289.94, sst_data[1], 2)
        self.assertAlmostEqual(289.89, sst_data[2], 2)
```

**Review by TonioF — CHANGES_REQUESTED** [review]
Thanks. Tests still did not run, suggestion for fix is in comment.

**Review by TonioF — APPROVED** [review]

**Pull request merged**